l1-contracts/contracts/bridge/L1SharedBridge.sol

```diff
@@ -417,47 +465,32 @@ contract L1SharedBridge is IL1SharedBridge, ReentrancyGuard, Ownable2StepUpgrade
             });
             require(proofValid, "yn");
         }
-        require(_amount > 0, "y1");
 
+        require(!_isEraLegacyDeposit(_chainId, _l2BatchNumber, _l2TxNumberInBatch), "ShB: legacy cFD");
         {
-            bool notCheckedInLegacyBridgeOrWeCanCheckDeposit;
-            {
-                // Deposits that happened before the upgrade cannot be checked here, they have to be claimed and checked in the legacyBridge
-                bool weCanCheckDepositHere = !_isEraLegacyDeposit(_chainId, _l2BatchNumber, _l2TxNumberInBatch);
-                // Double claims are not possible, as depositHappened is checked here for all except legacy deposits (which have to happen through the legacy bridge)
-                // Funds claimed before the update will still be recorded in the legacy bridge
-                // Note we double check NEW deposits if they are called from the legacy bridge
-                notCheckedInLegacyBridgeOrWeCanCheckDeposit = (!_checkedInLegacyBridge) || weCanCheckDepositHere;
-            }
-            if (notCheckedInLegacyBridgeOrWeCanCheckDeposit) {
-                bytes32 dataHash = depositHappened[_chainId][_l2TxHash];
-                bytes32 txDataHash = keccak256(abi.encode(_depositSender, _l1Token, _amount));
-                require(dataHash == txDataHash, "ShB: d.it not hap");
-                delete depositHappened[_chainId][_l2TxHash];
+            bytes32 dataHash = depositHappened[_chainId][_l2TxHash];
+            bytes32 assetId;
+            bytes32 txDataHash;
+            if (uint256(_assetId) <= type(uint160).max) {
+                // Claiming failed legacy deposit
+                assetId = keccak256(abi.encode(block.chainid, NATIVE_TOKEN_VAULT_VIRTUAL_ADDRESS, _assetId));
+                (uint256 _amount, ) = abi.decode(_assetData, (uint256, address));
+                txDataHash = keccak256(abi.encode(_depositSender, uint160(uint256(_assetId)), _amount));
```

Comment:
While it is more of a nit, I don't really like the fact we firstly encode these params, then decode them again. It would be better if we had some `internal` method that already accepts that `txDataHash` and the different legacy vs non-legacy logic was handled in another methods